internal/attestations/authorization.go

```diff
@@ -0,0 +1,174 @@
+	ErrAuthorizationNotFound = errors.New("requested authorization not found")
+)
+
+// Authorization is a lightweight record of a detached authorization in a gittuf
+// repository. It is meant to be used as a "predicate" in an in-toto
+// attestation.
+type Authorization struct {
+	TargetRef    string `json:"targetRef"`
+	FromTargetID string `json:"fromTargetID"`
+	ToTargetID   string `json:"toTargetID"`
+}
+
+// NewAuthorizationAttestation creates a new authorization for the provided
+// information. The authorization is embedded in an in-toto "statement" and
+// returned with the appropriate "predicate type" set.
+func NewAuthorizationAttestation(targetRef, fromTargetID, toTargetID string) (*ita.Statement, error) {
+	predicate := &Authorization{
+		TargetRef:    targetRef,
+		FromTargetID: fromTargetID,
+		ToTargetID:   toTargetID,
+	}
+
+	predicateBytes, err := json.Marshal(predicate)
+	if err != nil {
+		return nil, err
+	}
+
+	predicateInterface := &map[string]any{}
+	if err := json.Unmarshal(predicateBytes, predicateInterface); err != nil {
+		return nil, err
+	}
+
+	predicateStruct, err := structpb.NewStruct(*predicateInterface)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ita.Statement{
+		Type: ita.StatementTypeUri,
+		Subject: []*ita.ResourceDescriptor{
+			{
+				Digest: map[string]string{digestGitCommitKey: toTargetID},
+			},
+		},
+		PredicateType: AuthorizationPredicateType,
+		Predicate:     predicateStruct,
+	}, nil
+}
+
+// SetAuthorizationAttestation writes the new authorization attestation to the
+// object store and tracks it in the current attestations state.
+func (a *Attestations) SetAuthorizationAttestation(repo *git.Repository, env *sslibdsse.Envelope, refName, fromID, toID string) error {
+	if err := validateAuthorization(env, refName, fromID, toID); err != nil {
+		return err
+	}
+
+	envBytes, err := json.Marshal(env)
+	if err != nil {
+		return err
+	}
+
+	blobID, err := gitinterface.WriteBlob(repo, envBytes)
+	if err != nil {
+		return err
+	}
+
+	if a.authorizations == nil {
+		a.authorizations = map[string]plumbing.Hash{}
+	}
+
+	a.authorizations[AuthorizationPath(refName, fromID, toID)] = blobID
+	return nil
+}
+
+// RemoveAuthorizationAttestation removes a set authorization attestation
+// entirely. The object, however, isn't removed from the object store as prior
+// states may still need it.
+func (a *Attestations) RemoveAuthorizationAttestation(refName, fromID, toID string) error {
+	authPath := AuthorizationPath(refName, fromID, toID)
+	if _, has := a.authorizations[authPath]; !has {
+		return ErrAuthorizationNotFound
+	}
+
+	delete(a.authorizations, authPath)
+	return nil
+}
+
+// GetAuthorizationAttestationFor returns the requested authorization
+// attestation (with its signatures).
+func (a *Attestations) GetAuthorizationAttestationFor(repo *git.Repository, refName, fromID, toID string) (*sslibdsse.Envelope, error) {
```

Comment:
We probably need to tweak how we're storing attestations - I'd expect there to possibly be multiple attestations present for a given commit to/from pair. We can't rely on a single attestation per pair since these may need to be signed by different identities.

Reply:
I also don't know if we should be keying on ref, since I would expect "give me all attestations for this commit" to be a fairly common workflow - having to scan all refs for possible commits would be costly as the repo grows.

Reply:
> We probably need to tweak how we're storing attestations - I'd expect there to possibly be multiple attestations present for a given commit to/from pair. We can't rely on a single attestation per pair since these may need to be signed by different identities.

On this one, I disagree. In the `repository` package, the handler for adding an authorization for the same action when someone else signed one loads the attestation that exists and attaches a second signature to the DSSE envelope. That workflow isn't in this PR though. The predicate doesn't change, so that ought to be okay?

> I also don't know if we should be keying on ref, since I would expect "give me all attestations for this commit" to be a fairly common workflow - having to scan all refs for possible commits would be costly as the repo grows.

That's a good point and needs more thinking about. As a start I modeled this on the RSL reference entry as these are essentially additional signatures for an RSL. In the RSL we don't need to explicitly track the from ID, which is the only difference in the recorded information. I'll think about the data model some more.